f3/granite.go

```diff
@@ -511,39 +514,29 @@ func (i *instance) beginConverge() {
 	i.phaseTimeout = i.alarmAfterSynchrony(CONVERGE_PHASE.String())
 	prevRoundState := i.roundState(i.round - 1)
 
-	var justification Justification
+	// Proposal was updated at the end of COMMIT phase to be some value for which
+	// this node received a COMMIT message (bearing justification), if there were any.
+	// If there were none, there must have been a strong quorum for bottom instead.
+	var justification *Justification
 	if signers, signatures, ok := prevRoundState.committed.FindStrongQuorumAgreement(ZeroTipSetID()); ok {
-		value := ECChain{}
-		aggSignature := i.host.Aggregate(signatures, nil)
-		justificationPayload := Payload{
-			Instance: i.instanceID,
-			Round:    i.round - 1,
-			Step:     COMMIT_PHASE,
-			Value:    value,
-		}
-		justification = Justification{
-			Vote:      justificationPayload,
-			Signers:   signers,
-			Signature: aggSignature,
-		}
-	} else if signers, signatures, ok = prevRoundState.prepared.FindStrongQuorumAgreement(i.proposal.Head().CID); ok {
```

Comment:
This else if branch was searching for a quorum of received PREPARE messages to build a justification for the converge. But if this node has them, then it would have sent a message with them, and received its own message and thus stored them the same as if it received them from another node. So there's no need to do extra work for this case, just fall through to look up the received justifications.